gcloud-java-bigquery/src/test/java/com/google/gcloud/bigquery/DatasetTest.java

```diff
@@ -0,0 +1,226 @@
+  }
+
+  @Test
+  public void testReloadNull() throws Exception {
+    expect(bigquery.getDataset(DATASET_ID.dataset())).andReturn(null);
+    replay(bigquery);
+    assertNull(dataset.reload());
+  }
+
+  @Test
+  public void testReloadWithOptions() throws Exception {
+    DatasetInfo updatedInfo = DATASET_INFO.toBuilder().description("Description").build();
+    expect(bigquery.getDataset(DATASET_ID.dataset(), BigQuery.DatasetOption.fields()))
+        .andReturn(updatedInfo);
+    replay(bigquery);
+    Dataset updatedDataset = dataset.reload(BigQuery.DatasetOption.fields());
+    assertSame(bigquery, updatedDataset.bigquery());
+    assertEquals(updatedInfo, updatedDataset.info());
+  }
+
+  @Test
+  public void testUpdate() throws Exception {
+    DatasetInfo updatedInfo = DATASET_INFO.toBuilder().description("Description").build();
+    expect(bigquery.update(updatedInfo)).andReturn(updatedInfo);
+    replay(bigquery);
+    Dataset updatedDataset = dataset.update(updatedInfo);
+    assertSame(bigquery, dataset.bigquery());
+    assertEquals(updatedInfo, updatedDataset.info());
+  }
+
+  @Test
+  public void testDelete() throws Exception {
+    expect(bigquery.delete(DATASET_INFO.datasetId())).andReturn(true);
+    replay(bigquery);
+    assertTrue(dataset.delete());
+  }
+
+  @Test
+  public void testList() throws Exception {
+    BigQueryOptions bigqueryOptions = createStrictMock(BigQueryOptions.class);
+    PageImpl<BaseTableInfo> tableInfoPage = new PageImpl<>(null, "c", TABLE_INFO_RESULTS);
+    expect(bigquery.listTables(DATASET_INFO.datasetId())).andReturn(tableInfoPage);
+    expect(bigquery.options()).andReturn(bigqueryOptions);
+    expect(bigqueryOptions.service()).andReturn(bigquery);
+    replay(bigquery, bigqueryOptions);
+    Page<Table> tablePage = dataset.list();
+    Iterator<BaseTableInfo> tableInfoIterator = tableInfoPage.values().iterator();
+    Iterator<Table> tableIterator = tablePage.values().iterator();
+    while (tableInfoIterator.hasNext() && tableIterator.hasNext()) {
+      assertEquals(tableInfoIterator.next(), tableIterator.next().info());
+    }
+    assertFalse(tableInfoIterator.hasNext());
+    assertFalse(tableIterator.hasNext());
+    assertEquals(tableInfoPage.nextPageCursor(), tablePage.nextPageCursor());
+    verify(bigqueryOptions);
+  }
+
+  @Test
+  public void testGet() throws Exception {
+    BaseTableInfo info = TableInfo.builder(TableId.of("dataset", "table1"), Schema.of()).build();
+    expect(bigquery.getTable(TableId.of("dataset", "table1"))).andReturn(info);
+    replay(bigquery);
+    Table table = dataset.get("table1");
+    assertNotNull(table);
+    assertEquals(info, table.info());
+  }
+
+  @Test
+  public void testGetNull() throws Exception {
+    expect(bigquery.getTable(TableId.of("dataset", "table1"))).andReturn(null);
+    replay(bigquery);
+    assertNull(dataset.get("table1"));
+  }
+
+  @Test
+  public void testCreateTable() throws Exception {
```

Comment:
Do you think we should test the `create` functions that have `UserDefinedFunction` in their signature as well?  We could also add a `TableOption` as well.